### PR TITLE
Format arguments in userspace

### LINKED ISSUE #4: Refactor to do all formatting in userspace
Per discussion on Discord, we're interested in refactoring aya-log to something more [defmt-like][defmt], performing all formatting in userspace rather than using ufmt. I'm filing this as a tracking issue so we can discuss progress.

[defmt]: https://github.com/knurling-rs/defmt

### PR DESCRIPTION
**Pull request by vadorovsky**
This change moves away argument formatting from eBPF to the userspace.
eBPF part of aya-log writes unformatted log message and all arguments to
the perf buffer and the userspace part of aya-log is formatting the
message after receiving all arguments.

Aya-based project to test this change:

https://github.com/vadorovsky/aya-log-example

Fixes: #4
Signed-off-by: Michal Rostecki <vadorovsky@gmail.com>

### CODE REVIEW

**Comment by vadorovsky** [comment]
@alessandrod @Tuetuopay @noboruma PTAL

Everything we discussed on Discord should be addressed now.

File: ebpf/aya-log-ebpf-macros/src/expand.rs [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+                let t: Expr = input.parse()?;
+                input.parse::<Token![,]>()?;
+                Some(t)
+            }
+            false => None,
+        };
+
+        let lvl: Expr = input.parse()?;
+        input.parse::<Token![,]>()?;
+        let format_string: Expr = input.parse()?;
```
Comment (alessandrod) [inline ebpf/aya-log-ebpf-macros/src/expand.rs]: Shouldn't this be `LitStr`?

File: ebpf/aya-log-ebpf-macros/src/expand.rs [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+    syn::custom_keyword!(target);
+}
+
+impl Parse for LogArgs {
+    fn parse(input: ParseStream) -> Result<Self> {
+        let ctx: Expr = input.parse()?;
+        input.parse::<Token![,]>()?;
+
+        // Parse `target: &str`, which is an optional argument.
+        let target: Option<Expr> = match input.peek(kw::target) {
```
Comment (alessandrod) [inline ebpf/aya-log-ebpf-macros/src/expand.rs]: This should be an if

File: ebpf/aya-log-ebpf-macros/src/expand.rs [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+
+        // Parse variadic arguments.
+        let formatting_args: Option<Punctuated<Expr, Token![,]>> = if input.is_empty() {
+            None
+        } else {
+            input.parse::<Token![,]>()?;
+            Some(Punctuated::parse_terminated(input)?)
+        };
+
+        Ok(Self {
```
Comment (alessandrod) [inline ebpf/aya-log-ebpf-macros/src/expand.rs:59]: If you run cargo clippy, it'll tell you to fix a bunch of things here

File: ebpf/aya-log-ebpf-macros/src/expand.rs [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+                    #num_args,
+                ) {
+                    if let Ok(message_len) = ::aya_log_ebpf::write_record_message(
+                        &mut buf.buf[header_len..],
+                        #format_string
+                    ) {
+                        let record_len = header_len + message_len;
+
+                        if let Ok(record_len) = #quote_args {
+                            if record_len <= ::aya_log_ebpf::macro_support::LOG_BUF_CAPACITY {
```
Comment (alessandrod) [inline ebpf/aya-log-ebpf-macros/src/expand.rs]: Do we need this check? I think quote_args will `Err(())` if we run out of space

File: ebpf/aya-log-ebpf-macros/src/expand.rs [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    })
+}
+
+pub(crate) struct LogLvlArgs {
```
Comment (alessandrod) [inline ebpf/aya-log-ebpf-macros/src/expand.rs]: It seems a bit ugly to have these almost identical LogArgs and LogLvlArgs. I think they should either be merged somehow, or we should only have `LogLevelArgs` (two extra characters for readability :P), have `log(args: LogLevelArgs)` then parse the level, and pass it to `log_impl(level, args)`. `warn()` `error()` etc then would similarly do `log_impl(Level::Error, args)`, `log_impl(Level::Warn, args)` etc.

File: ebpf/aya-log-ebpf/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+impl<'a, T> TagLenValue<'a, T>
+where
+    T: Copy,
+{
+    #[inline(always)]
+    pub(crate) fn new(tag: T, value: &'a [u8]) -> TagLenValue<'a, T> {
+        TagLenValue { tag, value }
+    }
+
+    pub(crate) fn try_write(&self, mut buf: &mut [u8]) -> Result<usize, ()> {
```
Comment (alessandrod) [inline ebpf/aya-log-ebpf/src/lib.rs]: Maybe rename this to `write` for consistency

**Review by alessandrod — CHANGES_REQUESTED** [review]

File: ebpf/aya-log-ebpf-macros/src/expand.rs (hunk 5 above)
Comment (vadorovsky) [reply ebpf/aya-log-ebpf-macros/src/expand.rs]: Done, I managed to merge everything into `LevelArgs` and handle `level` as an optional argument.

**Comment by vadorovsky** [comment]
@alessandrod PTAL, your comments should be addressed

File: ebpf/aya-log-ebpf-macros/src/expand.rs [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+    }
+}
+
+pub(crate) fn log(args: LogArgs, level: Option<TokenStream>) -> Result<TokenStream> {
+    let ctx = args.ctx;
+    let target = match args.target {
+        Some(t) => quote! { #t },
+        None => quote! { module_path!() },
+    };
+    let lvl: TokenStream = if let Some(l) = args.level {
```
Comment (alessandrod) [inline ebpf/aya-log-ebpf-macros/src/expand.rs]: I think `level` should take precedence over `args.level`, or it becomes possible calling `error(Level::Trace, ...)` which seems wrong.

File: ebpf/aya-log-ebpf-macros/src/expand.rs [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+        quote! { #l }
+    } else if let Some(l) = level {
+        l
+    } else {
+        return Err(Error::new(
+            args.format_string.span(),
+            "`level` argument or calling other logging macros (i.e. error, info) is required",
```
Comment (alessandrod) [inline ebpf/aya-log-ebpf-macros/src/expand.rs]: I think this should say: "missing `level` argument: try passing a `aya_log_ebpf::Level` value"

File: ebpf/aya-log-ebpf-macros/src/expand.rs [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+        ));
+    };
+    let format_string = args.format_string;
+
+    let (num_args, quote_args) = match args.formatting_args {
```
Comment (alessandrod) [inline ebpf/aya-log-ebpf-macros/src/expand.rs]: nit: I think `quote_args` should be named `write_args`

**Review by alessandrod — CHANGES_REQUESTED** [review]
Great stuff! Only the `level` vs `args.level` thing then I think we're done!

File: ebpf/aya-log-ebpf-macros/src/expand.rs (hunk 8 above)
Comment (vadorovsky) [reply ebpf/aya-log-ebpf-macros/src/expand.rs]: Fun fact: this error message was fully generated by Copilot :smirk: But yes, I will fix it

File: ebpf/aya-log-ebpf-macros/src/expand.rs (hunk 9 above)
Comment (vadorovsky) [reply ebpf/aya-log-ebpf-macros/src/expand.rs]: done

File: ebpf/aya-log-ebpf-macros/src/expand.rs (hunk 7 above)
Comment (vadorovsky) [reply ebpf/aya-log-ebpf-macros/src/expand.rs]: done

**Review by alessandrod — APPROVED** [review]

**Comment by alessandrod** [comment]
Amazing work, thank you so much! 🙏🙏🙏

**Pull request merged**